raiden/transfer/channel.py

```diff
@@ -1310,6 +1312,9 @@ def events_for_close(
             None,
         )
 
+        # silence mypy: partner's balance proofs should be signed
+        assert not isinstance(channel_state.partner_state.balance_proof, BalanceProofUnsignedState)
```

Comment:
Hmm I am not sure about this one.
First if you wanted an assertion here, you should have a positive and not negative assertion. So `isinstance(channel_state.partner_state.balance_proof, BalanceProofSignedState)`.

But secondly when a `NettingChannelState` and a `NettingChannelEndState` is [initialized](https://github.com/raiden-network/raiden/blob/f5d48172931ce60c820253e81708acc2a2f49941/raiden/blockchain/state.py#L8) the initial `balance_proof` is `None`. The type of the balance proof considers `None` a valid value.

And it definitely is a valid value since the partner may have sent us no transfers. In which case when we handle the Channel close (see below) we sent empty hashes.

https://github.com/raiden-network/raiden/blob/f5d48172931ce60c820253e81708acc2a2f49941/raiden/raiden_event_handler.py#L241-L258

This probably also did not hit in any of the tests since the negative assertions holds for any type. e.g. 

```
True == not isinstance(str, BalanceProofUnsignedState)
True == not isinstance(None, BalanceProofUnsignedState)
```
e.t.c.

P.S.: That means that `ContractSendChannelClose` constructor also need the `balance_proof` argument to be `typing.Optional[BalanceProofSignedState]` and not `BalanceProofSignedState`.

Reply:
That's why I wanted somebody's eye on this one.

I'll change the assertion to positive, but it has to allow `None` in addition to `BalanceProofSignedState`.

> P.S.: That means that ContractSendChannelClose constructor also need the balance_proof argument to be typing.Optional[BalanceProofSignedState]

Yes, mypy forced me to make that change (already in this PR)
https://github.com/raiden-network/raiden/pull/3214/files#diff-dbf5b17f182eaefb153df563cd0418deR27

Reply:
Done.  Is this what you wanted? https://github.com/raiden-network/raiden/pull/3214/files#diff-fe7772da6084a3493b31e88fbf7f6d55R1317 @LefterisJP

Reply:
yes thank you.